README.rst
dipplanner Quick Start
======================

.. warning::

    This software is provided free of charge for experienced divers to evaluate diving profiles.
    This software is highly experimental, it probably contains bugs and should not be relied upon for actual dives. Use it at your own risk.

    If you follow dive schedules generated by this software you could suffer decompression sickness, serious injury or possibly death.
    The author does not warrant that this software accurately reflects Buhlmann's algorithms, or that it will produce safe, reliable, results.

    Diving in general is fraught with risk, and decompression diving using mixed gases and rebreathers adds significantly more risk. This software is not intended for uneducated users.
    This software and the decompression schedules it produces are tools for experienced divers only.

    IF YOU DO NOT UNDERSTAND OR DO NOT AGREE TO THIS STATEMENT DO NOT USE THIS SOFTWARE

Description
-----------

Model supported
^^^^^^^^^^^^^^^

* Bülhmann ZH-L16B with Gradient Factor
* Bülhmann ZH-L16C with Gradient Factor

Gases
^^^^^

any mix of oxygen, helium and nitrogen is supported, this include:

* Air
* Nitrox
* Oxygen
* Trimix
* Heliox
* ...

Other Features
^^^^^^^^^^^^^^

* Open Circuit Dives
* CCR Dives

  * setpoint
  * bailout

* Tank planification

  * Tank size and pressure
  * Simple or Double Tank
  * gaz consumption rate
  * rules for warning (rules of third, rules in remaining pressure, etc..)
  * automatic (or non automatic) refill of Tank between successive dives

* Multigaz (multi Tank)
* Automatic selection of best gaz
* Multilevel Dives
* Altitude Dives
* Successive Dives
* No Flight Time Calculation
* CNS Tracking
* OTU Tracking
* MOD Calculation
* END Calculation
* Adjustable gradient factor and stop depth level

Calculations
^^^^^^^^^^^^

The main principe of dipplanner calculation are to be as accurate as possible,
using the most precise algorithm.

* All internal calculation in SI Units
* Most precise calculations (trying to avoid any approximation):

  * second to second decompression time calculation (adjustable)
  * real water density calculation
  * real gaz pressure calculation in tanks using van der waals
  * real surface pressure calculation (using Antoine equation for ppH2O)
  * ...

* Every parameter is adjustable

.. warning:: in result of the "precise" calculations, the planned dives may be
             shorter than other equivalent dive planification programs.

             Users of dipplanner MUST be aware of this difference and adjust
             the conservatisms to their wishes.

             On the other way, with the most precise calculations, dipplanner
             try to avoid unexpected effects (or worse: unseen effects) of
             approximations. (approximation effect to conservatism are sometimes
             counter-intuitive and may in some cases reduce the conservatism
             without warning)

Get the sources
---------------

latest source code (unstable)

please use git to clone the repository :

::

    git clone git://github.com/ThomasChiroux/dipplanner.git

Preparation: virtualvenv
^^^^^^^^^^^^^^^^^^^^^^^^

It is recommended (but optional) to setup a virtualvenv before running or installing.
In this way, you will not break or change your real environment.
So after getting the source, go in the dipplanner directory:

dipplanner needs at least python 3.4 to work.

::

   cd dipplanner
   mkdir venv
   virtualenv venv
   source venv/bin/activate

Installing
----------

Inside the dipplanner source directory, run:

::

    python setup.py install

Using
-----

This version if currently only usable in command line

Run the program
^^^^^^^^^^^^^^^

ex:

::

    dipplanner --help


Planning one dive
^^^^^^^^^^^^^^^^^

To plan a dive, you should at least provide one tank and one dive segment.
Here is below a sample for a 12l tank with 200b of air and a dive of 25min at 30m

::

    dipplanner -t "airtank;0.21;0.0;12;200;50b" -s "30;25*60;airtank;0.0"

You can provide more than one tank and of course multiple segments (they will be processed in the order you provided it)
Deco tanks will be automaticaly choosen if appropriate.
Here is below a sample for a trimix dive : bi 12l-cylinder of Tx21/30 and Deco Nx80 (S80), 50m - 20mins:

::

    dipplanner -t "tx;0.21;0.30;24;200;50b" -t "deco;0.8;0.0;12;200;50b" -s "50;20*60;tx;0.0"

Change some parameters of the dive
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

See ::

    dipplanner --help

output to see all available parameters

You can also read :ref:`dipplanner_cmdline`

Config files
^^^^^^^^^^^^

You can provide one or several config files to the program.
The config file can overide any default parameter.
see config/default_config.cfg for all the details

parameter are changed using this order:

1) default parameter
2) parameter set in config files
3) parameter set in command line

You can also read :ref:`dipplanner_configfile`

Units : SI or Imperial
----------------------

dipplanner uses only SI unit internally.
However a config parameter or a command line parameter can change this behaviour.
and use imperial Units

if imperial unit is set in a config file :
   all the parameters from this config file and all the config files
   read after will be considered imperial (including command line parameters)
   But all the parameter in previous config files will still be considered as SI
   All the ouput will be done in imperial units

if imperial unit is set in command line :
   all the parameter given in command line will be considered imperial,
   but not the parameters eventually given using config files
   All the output will be done in imperial units

SI and imperial unit converter uses the following correspondances:

* bar <--> psi
* liter <--> cubic feet
* meter <--> feet

References
----------

* at first this program is a python rewrite of MV-Plan a dive planning tool written in java by Guy Wittig
* `ref used for programming and understand algorithms <https://github.com/ThomasChiroux/dipplanner/wiki/Reference-used-for-programming-and-understand-algorithms/>`_.

Open Source and Licence
-----------------------

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.
If not, see http://www.gnu.org/licenses/gpl.html


NEWS.rst
Release Notes
=============

v0.3 (ongoing)
--------------

.. note::
    v0.3 is not released yet.

    Below are the feature already developed for the 0.3 target release.

    See GitHub Issues for more infos.

Python
******

* dipplanner is now (only) python 3.4+ compatible
  now old python 2.7 compatible version is frozen in py2.7 branch.

New features
************

* No-flight time calculation
* non-blocking dive situations errors

Bug corrections
***************

* tank infos on repetitive dives was wrong (only last tank status was
  diven, even for the firs(s) dive(s).

v0.2
----

New features
************

* New submodel: buhlmann ZH-L16C
* Handle repetitive dives
* Variable ppH2O in surface based on Temp and % humidity
* Limit CCR dive depth based on diluant values
* Export dive plannification in different format
* Config Files
* Tanks:
    * More accurate calculation of tank volume
    * minimum gas remaining rules in tanks
    * handle double tanks

Bug corrections
***************

* Automatic gas selection was not well handled on Hypoxic trimix dives

v0.1
----

New features
************

* Same 'base' functionnality as MVPlan (v1.5):
    * buhlmann ZH-L16B
    * mv-value gradient conservatism (adjustable)
    * oxygen toxicity (OTU and CNS) calculation
    * support Open Circuit dives : air, nitrox, trimix, heliox...
    * support CCR dives with OC deco/bailout if wanted
    * support dive in altitude
    * automatic selection of deco when ascending
    * ...
* But some differences:
    * command line only (or direct use in python) : no GUI, no automatic update
    * only in english
    * the unit system is only SI:
        * meter
        * seconds
        * bar (almost SI: 1 bar = 10E5 pascal)
        * liter (dm3)
    * Uses 'Tanks' instead of 'Gases' in order to calculate gas consumption for each tank and raise error when dive reach empty tank
    * All time parameters and calculations are by default in seconds instead of minutes
    * the END 'Equivalent Narcosis depth' is calculated based on narcosis index of every breathed gases (N2, He and O2) instead of the 'Only N2 is narcotic' in MVPlan
    * water pressure is by default calculated using the 'real pressure' of water (wich can change between fresh or sea water) instead of '10m == 1bar'


settings.py
#! /usr/bin/python
# -*- coding: utf-8 -*-
#
# Copyright 2011 Thomas Chiroux
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.
# If not, see <http://www.gnu.org/licenses/gpl.html>
#
# This module is part of dipplanner, a Dive planning Tool written in python
"""Global settings for dipplanner
"""

__authors__ = [
    # alphabetical order by last name
    'Thomas Chiroux', ]

# ======= "Internal" Settings ========
# software version, populated automatically
__VERSION__ = None

# unless knowing what you're doing, this prefs should not be changed
# by the user
# water density kg/l
FRESH_WATER_DENSITY = 1.0
SEA_WATER_DENSITY = 1.03
ABSOLUTE_MAX_PPO2 = 2.0
ABSOLUTE_MIN_PPO2 = 0.16
ABSOLUTE_MAX_TANK_PRESSURE = 300  # in bar
ABSOLUTE_MAX_TANK_SIZE = 2 * 20  # in liter

# Temperature at surface.
# Used to calculate PP_H2O_SURFACE
SURFACE_TEMP = 20

HE_NARCOTIC_VALUE = 0.23
N2_NARCOTIC_VALUE = 1.0
O2_NARCOTIC_VALUE = 1.0
AR_NARCOTIC_VALUE = 2.33

STOP_DEPTH_INCREMENT = 3  # in meter
LAST_STOP_DEPTH = 3  # in meter : last stop before surfacing
STOP_TIME_INCREMENT = 1  # in second
FORCE_ALL_STOPS = True  # one deco stop begun, force to stop to each deco
                        # depth stop
AMBIANT_PRESSURE_SEA_LEVEL = 1.01325  # surface pressure (in bar)
METHOD_FOR_DEPTH_CALCULATION = 'complex'  # either simple (/10) or complex
TRAVEL_SWITCH = 'late'  # "late" or "early"

# #TODO: put this also in config.cfg
FLIGHT_ALTITUDE = 2450  # in meter practicly, this value can not be bigger than
                        # about 2850m because breathing air at 0m will only
                        # 'prepare the body' to a decompression until this
                        # value. To go higher, another 'stop' is needed
                        # between.

# ========= User settings ========
TEMPLATE = "default-color.tpl"

DECO_MODEL = "ZHL16c"
WATER_DENSITY = SEA_WATER_DENSITY
AMBIANT_PRESSURE_SURFACE = AMBIANT_PRESSURE_SEA_LEVEL

DEFAULT_MAX_PPO2 = 1.6
DEFAULT_MIN_PPO2 = 0.21
DEFAULT_MAX_END = 30  # in meter

DIVE_CONSUMPTION_RATE = 17.0 / 60  # liter/s
DECO_CONSUMPTION_RATE = 12.0 / 60  # liter/s

DESCENT_RATE = float(20) / 60  # m/s
ASCENT_RATE = float(10) / 60  # m/s

# Warning : if RUN-TIME is True, the segment duration must include descent time
# if the duration is too small dipplanner will raise an error
RUN_TIME = True  # if True: segments represents runtime,
                 # if false, segments represents segtime

USE_OC_DECO = True  # if True, use enabled gases of decomp in oc or bailout

GF_LOW = 0.30
GF_HIGH = 0.80

# TODO: check the usage of the MULTILEVEL_MODE setting
MULTILEVEL_MODE = False

AUTOMATIC_TANK_REFILL = True
